### PR TITLE
net-dns/letsdns: New package

### PR DESCRIPTION
**Pull request by rseichter**
Manage DANE TLSA records in DNS servers.

The ebuild deletes the "tests" directory to avoid the following build-time error:

> Package installs 'tests' package which is forbidden and likely a bug in the build system.

As the author of LetsDNS, I can rely on the fact that my CI tests are successful.

### CODE REVIEW

**Comment by Flowdalic** [comment]
Thanks for the contribution. That looks like an interesting package.

> > Package installs 'tests' package which is forbidden and likely a bug in the build system.
> 
> As the author of LetsDNS, I can rely on the fact that my CI tests are successful.

Note that this QA warning indicates a faulty build system configuration. See also https://projects.gentoo.org/python/guide/distutils.html#preventing-test-directory-from-being-installed. You as upstream author should be able to easily fix this.

**Comment by rseichter** [comment]
> Note that this QA warning indicates a faulty build system configuration. [...] You as upstream author should be able to easily fix this.

I'm not sure I understand your meaning. How could my configuration be faulty when I can build distribution packages by running `python -m build` and nothing else, for Python versions from 3.7 to 3.10, for both manual and CI builds?

File: net-dns/letsdns/letsdns-1.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+DESCRIPTION="Manage DANE TLSA records in DNS servers"
+HOMEPAGE="https://letsdns.org"
+SRC_URI="https://github.com/LetsDNS/letsdns/archive/refs/tags/${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="GPL-3+"
+SLOT="0"
+KEYWORDS="~amd64"
+
+DISTUTILS_USE_SETUPTOOLS="pyproject.toml"
```
[the remark this answers is not in the document]
Comment (rseichter) [reply net-dns/letsdns/letsdns-1.0.ebuild:18]: Fixed.

**Comment by rseichter** [comment]
I figured out what you mean. Given that the release tarball is frozen, I have opted to patch setup.cfg in the ebuild. The tests which fail now are not critical and are caused by differences between my CI environment and the Gentoo ebuild sandbox (availability of files/commands, DNS servers). I don't think I can clean this up further without making changes upstream, but many tests are skipped anyway. Better to run some than none, I guess.

**Comment by Flowdalic** [comment]
nit: The commit message still talks about the `tests/` directory being deleted.

File: net-dns/letsdns/letsdns-1.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	dev-python/requests[${PYTHON_USEDEP}]"
+
+python_test() {
+	export LOG_LEVEL=FATAL
+	export UNITTEST_CONF=tests/citest.conf
+	"${EPYTHON}" -m unittest discover
+}
+
+python_prepare_all() {
+	echo -e '[options.packages.find]\nexclude=tests*' >>setup.cfg || die
```
Comment (Flowdalic) [inline net-dns/letsdns/letsdns-1.0.ebuild:29]: This should be done in `src_prepare`.

**Comment by rseichter** [comment]
I moved code to src_prepare() and amended the commit message.

**Comment by Flowdalic** [comment]
> I moved code to src_prepare()

I am sorry, that was me not looking correctly at the ebuild (I did not see that this already used `python_prepare`). I think that https://github.com/gentoo/gentoo/commit/b5a3989d484fb58d823feb108bddb70ef213c034#diff-1d6b14570a83e37d1bb0a8db5632acae4305b258132e56f2c4086fc6c17eee35R28-R31 is actually the right way. Hence I deleted my comment immediately.

File: net-dns/letsdns/letsdns-1.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+SLOT="0"
+KEYWORDS="~amd64"
+
+RDEPEND="dev-python/cryptography[${PYTHON_USEDEP}]
+	dev-python/dnspython[${PYTHON_USEDEP}]
+	dev-python/requests[${PYTHON_USEDEP}]"
+
+src_prepare() {
+	eapply_user
+	echo -e '[options.packages.find]\nexclude=tests*' >>setup.cfg || die
```
Comment (Flowdalic) [inline net-dns/letsdns/letsdns-1.0.ebuild:24]: Please do not forget to also fix this upstream at https://github.com/LetsDNS/letsdns/blob/master/setup.cfg

Comment (rseichter) [reply net-dns/letsdns/letsdns-1.0.ebuild:24]: "There's a brach for this." 😉

**Comment by rseichter** [comment]
> I am sorry, that was me not looking correctly at the ebuild

I moved things back to python_prepare_all(), and I hope that's this ebuild completed? Pretty please.

**Comment by Flowdalic** [comment]
Running `ebuild letsdns-1.0.ebuild test` fails for me here:

```
>>> Test phase: net-dns/letsdns-1.0
 * python3_9: running distutils-r1_run_phase python_test
.WARNING:root:FailExec.execute returned code -20
.WARNING:root:FailPost.teardown returned code -30
.WARNING:root:FailPre.setup returned code -10
......sEdiff: dump-expected: No such file or directory
F.....ssssss....ss......
======================================================================
ERROR: test_lifecycle (tests.test_action.NsupdateTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data-scratch/var-tmp/portage/net-dns/letsdns-1.0/work/letsdns-1.0/tests/test_action.py", line 113, in test_lifecycle
    rc = NsupdateStdout.lifecycle(self.c, NsupdateStdout())
  File "/data-scratch/var-tmp/portage/net-dns/letsdns-1.0/work/letsdns-1.0/letsdns/nsupdate.py", line 24, in lifecycle
    return rdata_action_lifecycle(conf, action)
  File "/data-scratch/var-tmp/portage/net-dns/letsdns-1.0/work/letsdns-1.0/letsdns/tlsa.py", line 65, in rdata_action_lifecycle
    records = tlsa_records(conf)
  File "/data-scratch/var-tmp/portage/net-dns/letsdns-1.0/work/letsdns-1.0/letsdns/tlsa.py", line 56, in tlsa_records
    cert = read_x509_cert(path)
  File "/data-scratch/var-tmp/portage/net-dns/letsdns-1.0/work/letsdns-1.0/letsdns/crypto.py", line 29, in read_x509_cert
    with open(filename, 'rb') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'citest.crt'

======================================================================
FAIL: test_dump (tests.test_config.ConfigurationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data-scratch/var-tmp/portage/net-dns/letsdns-1.0/work/letsdns-1.0/tests/test_config.py", line 44, in test_dump
    self.assertEqual(0, diff.returncode)
AssertionError: 0 != 2

----------------------------------------------------------------------
Ran 34 tests in 3.537s

FAILED (failures=1, errors=1, skipped=9)
>>> Completed testing net-dns/letsdns-1.0
```

**Comment by rseichter** [comment]
The failing tests are expected, see https://github.com/gentoo/gentoo/pull/25011#issuecomment-1097118597 .

**Comment by Flowdalic** [comment]
> The failing tests are expected, see https://github.com/gentoo/gentoo/pull/25011#issuecomment-1097118597 .

I see. I somehow assume that `export UNITTEST_CONF=tests/citest.conf` would enable the working unit tests. Then you may want to add `RESTRICT=test` for the time being. And I suggest to split the tests into tests that should success when run in `src_test`, and the ones that would probably not work, e.g., tests that require network connections, and only enable the former in `src_test` while removing the test restriction.

**Comment by rseichter** [comment]
Don't get me wrong, but the tests are written to work in both my develpment- and CI-environment, which they do just fine. I don't write tests to satisfy Gentoo ebuild sandboxes. While I might revisit the tests some time in the future, I won't change anything else right now. We can either accept the failing tests as they are, or I can remove the tests altogether, as I did in my very first version of the ebuild.

**Comment by rseichter** [comment]
Do you want me to remove the tests or leave the ebuild as it is right now?

**Comment by thesamesam** [comment]
I'd like you to get a reasonable subset working if not the full suite.

We don't even take new packages from users right now unless there's a good reason to, a developer is interested in it, or it's been popular in e.g GURU.

Test suites are important to know things work correctly in Gentoo. Working in upstream CI isn't the same thing. I spend a significant amount of my time working on test issues in packages. It's part of the job.

**Comment by Flowdalic** [comment]
> a Gentoo ebuild is a mere bonus, and I can use a personal overlay for that

Note that submitting the package to ::guru is also an option.

**Comment by rseichter** [comment]
I am aware of GURU, thanks. What this pull request apparently boils down to is one of the following:

1. You accept an ebuild version without any tests.
2. You accept that three non-critical tests fail in the Gentoo ebuild sandbox (and only there!) for ebuild version 1.0.
3. You are troopers and patch out these three tests yourselves, because I simply don't have the time, and I keep Gentoo peculiarities in mind for future LetsDNS releases.
4. You don't accept the pull request.

**Comment by thesamesam** [comment]
I don't think they're Gentoo peculiarities. It's quite common to want tests to pass in a network sandbox. It's not actually specific to some sort of Gentoo _problem_. We expect ebuilds submitted to ::gentoo to meet our standards, which includes running a test suite if it exists and it's possible.

Your #3 point is rather rude -- you're submitting something to us, we're not obligated to accept it, but we've spent time reviewing it and explained the issues. If you don't have the time to maintain it, I'm not sure why you're submitting it.

Ebuilds being in ::gentoo is not a right. You can put them in an overlay for now if you're not willing to meet our QA standarsd. If you're happy to meet the standards we've explained in this PR, feel free to reopen, but it doesn't sound like the case right now.

**Comment by rseichter** [comment]
Point number three was merely meant as a confirmation that, should you so choose, make the changes you deem necessary without me raising any objection. I won't have the time myself because of family time the Easter weekend. We can wait for the next LetsDNS release, but there is no scheduled date yet.

**Pull request closed without merging**